packages/insured-bridge-relayer/src/ProfitabilityCalculator.ts

```diff
@@ -0,0 +1,236 @@
+  ) {
+    this.relayerDiscount = toBNWei(relayerDiscountNumber).div(toBN("100"));
+    if (this.relayerDiscount.lt(toBN(0)) || this.relayerDiscount.gt(toBNWei(1)))
+      throw new Error("Bad discount factor! 0<=relayerDiscount<=1e18");
+
+    this.coingecko = new Coingecko();
+  }
+
+  async update() {
+    this.logger.debug({ at: "ProfitabilityCalculator", message: "Updating prices", tokenList: this.l1Tokens });
+
+    // If the l1TokenInfo is empty then this is the first run. Figure out the token types of each l1Token.
+    if (Object.keys(this.l1TokenInfo).length == 0) {
+      const [umaAddress, wethAddress] = await Promise.all([
+        await getAddress("VotingToken", this.l1ChainId),
+        await getAddress("WETH9", this.l1ChainId),
+      ]);
+      for (const l1Token of this.l1Tokens) {
+        this.l1TokenInfo[l1Token] = { tokenType: TokenType.ERC20, tokenEthPrice: toBNWei("1") };
+        if (l1Token == umaAddress) this.l1TokenInfo[l1Token].tokenType = TokenType.UMA;
+        else if (l1Token == wethAddress) this.l1TokenInfo[l1Token].tokenType = TokenType.WETH;
+        else this.l1TokenInfo[l1Token].tokenType = TokenType.ERC20;
+      }
+    }
+
+    // Fetch the prices of each token, denominated in ETH. If coingecko does not have the price or is down then this
+    // call will fail and the tokenPrices array will have this error.
+    const tokenPrices = await Promise.allSettled(
+      this.l1Tokens.map((l1Token) => this.coingecko.getCurrentPriceByContract(l1Token, "eth"))
+    );
+
+    // For each token, extract the price and convert to a wei'd BN. Note that if the type is WETH then the price is 1.
+    // This is done as coingecko does not always return the price of 1 for WETH, in ETH. Note that if the status is not
+    // fulfilled then the call failed. In this case we dont have pricing information for this token. Set price to 0 to
+    // not send arrays (always will be unprofitable) and generate a warning.
+    for (const [index, priceResponse] of tokenPrices.entries()) {
+      if (priceResponse.status == "fulfilled") {
+        this.l1TokenInfo[this.l1Tokens[index]].tokenEthPrice =
+          this.l1TokenInfo[this.l1Tokens[index]].tokenType == TokenType.WETH
+            ? toBNWei("1")
+            : toBNWei(priceResponse.value[1]);
+      } else {
+        this.l1TokenInfo[this.l1Tokens[index]].tokenEthPrice = toBNWei("0");
+        this.logger.warn({
+          at: "ProfitabilityCalculator",
+          message: "Could not find token price!",
+          token: this.l1Tokens[index],
+        });
+      }
+    }
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Updated prices",
+      tokenInfo: objectMap(this.l1TokenInfo, (value: { tokenType: TokenType; tokenEthPrice: BN }) => {
+        return { tokenType: TokenType[value.tokenType], tokenEthPrice: fromWei(value.tokenEthPrice) };
+      }),
+    });
+  }
+
+  getRelaySubmitTypeBasedOnProfitability(
```

Comment:
Is this the method that users call?

I could totally be wrong here, but if I were a caller, I would prefer for this class to _just_ tell me how profitable each action is and then I could decide which one to do.

For instance, what happens if the relay has already been slow relayed and _all_ I can do is speed it up. If this module told me to send an instant relay, that doesn't tell me anything about the profitability of a speed up, right? Seems like the caller might have other constraints to consider along with profitability that this module can't take into account.

Thoughts?

Reply:
ye, it is a tradeoff on where that decision-making happens. this chunk of logic at the end here ( from line 134 to line 142) could easily sit in the `Relater.ts` code block (in fact that part of code was originally sitting in the relayer and I moved it here!). 

my thinking in this was that all decision-making is informed exclusively by profit, in all cases. in your example where the relay has already been slow relayed and all that can be done is sped up then the inputs to this method would have `speedUpRevenue` and `instantRevenue` set to zero so the only profitable action that would be returned would be `RelaySubmitType.SpeedUp` (note there a number of unit tests that show this particular case (as well as others) in the `ProfitabiliyCalculator.ts` tests).

At the end of the day, it's just a question of where the logic sits. the assumption is the caller has already done a fair amount of logic on the revenues for each action according to what _is_ possible and that influences the profit computation.

You do have a good argument on returning profit from this module then doing the casework within the client. I'm happy to move it there if you feel strongly about this.

Reply:
I see. That makes sense. A few quick points:
- Seems a little convoluted to have the relayer signal impossible cases by sending 0. But it does make sense.
- If the relayer never sees profitability, this limits our flexibility to do things with that information, like log the details in an error (since we would probably want to understand that we chose not to do the relay and by what margin was that decision made).

The latter bullet is the one I care about the most, since if we really are seeing an influx of underpriced txns, we'll need to have well-tailored errors and logs to tell us what's happening and allow us to potentially override the bot's decisions. Do you think that's possible in the current setup? If so, I'm not strongly opinionated about changing this structure.

Reply:
So regarding logs the way it's structured right now I actually thing provides far, far more detailed logs (this is part of the reason why I chose this pattern). There are two main logs that are produced during the profitability evaluation:
a) `Considering relay profitability` that is produced at the beginning and  b) `Considered relay profitability` which is produced at the end. BOth of these logs contain _all_ state data the `ProfitabilityCalculator` has on this particular token including. The first log contains:
```ts
	  l1Token,
      tokenType: TokenType[tokenType],
      tokenEthPrice: fromWei(tokenEthPrice),
      cumulativeGasPrice: cumulativeGasPrice.toString(),
      relayerDiscount: fromWei(this.relayerDiscount),
      slowRevenue: slowRevenue.toString(),
      speedUpRevenue: speedUpRevenue.toString(),
      instantRevenue: instantRevenue.toString(),
```

and the second log contains:
```ts
      l1Token,
      tokenType: TokenType[tokenType],
      tokenEthPrice: fromWei(tokenEthPrice),
      cumulativeGasPrice: cumulativeGasPrice.toString(),
      relayerDiscount: fromWei(this.relayerDiscount),
      ethSubmissionCost: objectMap(ethSubmissionCost, (value: BN) => fromWei(value)),
      ethRevenue: objectMap(ethRevenue, (value: BN) => fromWei(value)),
      ethProfitability: objectMap(ethProfitability, (value: BN) => fromWei(value)),
      relaySubmitType: RelaySubmitType[relaySubmitType],
```

This is far more info than what would be possible if this decision making happened within the relayer as the relayer would not have information about: 1) tokenEthPrice, 2) relayerDiscount, 3) cumulativeGasPrice or 4) ethSubmissionCost. 

Put differently, by keeping this logic here _all_ information is processed in the same place enabling for more detailed logs to help with debugging.

Reply:
But these are all debug logs, right? We won't be able to relay this info to slack unless we have _context_ about whether the decision is notable enough that it needs to be escalated.

That context is what I think may be missing in the profitability calculator.

Reply:
ye, you right. these logs are solely sent as info in the profitability calculator. this means that once the call has left the profitability calculator this context is lost and all we have in the `Relayer` was the _outcome_ and no info on why that was made (that was sent as an info log in the Profitability calculator). I see your point about wanting that context in the Relayer, in particular, to inform the log sent when a relay is deemed unprofitable such as `Not relaying potentially unprofitable deposit, or insufficient balance 😖`. it might be useful to have the additional profitability logic in that log showing the costs.

however, to get that to work cleanly I think we'll need to refactor some of the flow in the `Relayer.ts` file to more consistently pass this information around.

I propose for now we leave it as this PR has it and make an overall refactor down the line to better pass around this data. this PR is already almost 800 lines and adding in better data preservation for this component will be pretty considerable (likely over 200 lines). is that ok for you?

Reply:
Seems reasonable -- I don't think it's a huge deal for now as the relayer can at least tell us when it ignores something if it wants to. It just won't be able to give us the numbers via slack.